libraries/AP_HAL_Linux/RCInput_RPI.cpp

```diff
@@ -524,8 +524,18 @@ void RCInput_RPI::_timer_tick()
         return;
     }
 
+    const uint32_t offset = con_blocks->get_offset(con_blocks->_virt_pages,(uintptr_t)ad);
     for (int j = 1; j >= -1; j--) {
-        void *x = circle_buffer->get_virt_addr((ad + j)->dst);
+        
+        // Get address of next or previous (dma_cb_t)
+        ad = (dma_cb_t *)con_blocks->get_page(con_blocks->_virt_pages,offset + (uint32_t)(sizeof(dma_cb_t) * j));
+        if (!ad) 
+        {
+           continue ;
```

Comment:
What I don't understand here is why this would ever be NULL. We don't crash now, which is good, but we may also be hiding another problem.

Reply:
`dma_cb_t *ad = (dma_cb_t *)con_blocks->get_virt_addr(dma_reg[RCIN_RPI_DMA_CONBLK_AD | RCIN_RPI_DMA_CHANNEL << 8]);`

if **ad is not** null then it is a valid address for sure. that is true.

However, in later lines of code we add `(sizeof(dma_cb_t) * j)` to **ad** original address which may point out of range, as DMA could be returning the last **dma_cb_t** address.
The same case when you subtract  `(sizeof(dma_cb_t) * j)` from **ad** original address if DMA points to the first **dma_cb_t** address.

i.e. condition in **Memory_table::get_page**
`(addr >= PAGE_SIZE * _page_count)` returns true

So this is a sanity check.

Reply:
Please let me know your feedback on my reply. 

The code is not crashing and RCInput_RPI works very well in all my trials. and there is no bug appears.
Potential bugs could be everywhere and that is why we keep fix it :). Right now the code has crash bug solved and it is no more crashing anymore.
Also I used **con_blocks->get_page** the same way used in **init_ctrl_data()** actually this is the reason I discovered the original cause of the issue. 
**init_ctrl_data** does not use if for null because it already knows where the pointer exactly for the **_timer_tick** function we dont know which block we are pointing to.